ray-curator/ray_curator/stages/deduplication/fuzzy/minhash.py

```diff
@@ -0,0 +1,326 @@
+        Width of character n-grams for minhashing
+    num_hashes : int, default=260
+        Number of hash functions (length of minhash signature)
+    seed : int, default=42
+        Random seed for reproducible minhash generation
+    use_64bit_hash : bool, default=False
+        Whether to use 64-bit hash functions (vs 32-bit)
+    read_format : Literal["jsonl", "parquet"], default="jsonl"
+        Format of input files
+    read_kwargs : dict[str, Any] | None, default=None
+        Additional keyword arguments for reading input files
+    write_kwargs : dict[str, Any] | None, default=None
+        Additional keyword arguments for writing output files
+
+    Examples
+    --------
+    >>> stage = MinHashStage(
+    ...     output_dir="/path/to/minhash/output",
+    ...     text_column="content",
+    ...     num_hashes=128,
+    ...     char_ngrams=5
+    ... )
+    >>> # Use in a pipeline to process document batches
+    """
+
+    def __init__(  # noqa: PLR0913
+        self,
+        output_dir: str,
+        text_column: str = "text",
+        minhash_column: str = "_minhash_signature",
+        char_ngrams: int = 24,
+        num_hashes: int = 260,
+        seed: int = 42,
+        use_64bit_hash: bool = False,
+        read_format: Literal["jsonl", "parquet"] = "jsonl",
+        read_kwargs: dict[str, Any] | None = None,
+        write_kwargs: dict[str, Any] | None = None,
+        pool: bool = True,
+    ):
+        # Initialize parent classes
+        ProcessingStage.__init__(self)
+        DeduplicationIO.__init__(self, id_generator=None)
+
+        # Set ProcessingStage attributes
+        self._name = self.__class__.__name__
+        self._resources = Resources(gpus=1.0)  # Requires 1 GPU
+
+        self.output_dir = output_dir
+        self.text_column = text_column
+        self.minhash_column = minhash_column
+        self.char_ngrams = char_ngrams
+        self.num_hashes = num_hashes
+        self.seed = seed
+        self.use_64bit_hash = use_64bit_hash
+        self.read_format = read_format
+        self.read_kwargs = read_kwargs or {}
+        self.write_kwargs = write_kwargs or {}
+        self.pool = pool
+        # Initialize the minhash processor in setup
+        self.minhash_processor = None
+        self.id_generator = None
+
+    def setup(self, _worker_metadata: "WorkerMetadata | None" = None) -> None:
+        """Initialize the GPU MinHash processor and ID generator."""
+        # Initialize the ID generator (will be shared across workers)
+        self.id_generator = IdGenerator.options(
+            name=CURATOR_ID_GENERATOR_ACTOR_NAME, get_if_exists=True, lifetime="detached"
+        ).remote()
+
+        # Initialize the GPU minhash processor
+        self.minhash_processor = GPUMinHash(
+            seed=self.seed,
+            num_hashes=self.num_hashes,
+            char_ngrams=self.char_ngrams,
+            use_64bit_hash=self.use_64bit_hash,
+            pool=self.pool,
+        )
```

Comment:
Out of curioisity, wondering why this is as a design choice of splitting stage and processor separately rather than keeping everything inside a `MinhashStage`?

Reply:
Probably the same as this: https://github.com/NVIDIA-NeMo/Curator/pull/837#discussion_r2261207607